src/swarm-tui/Swarm/TUI/Model/Event.hs

```diff
@@ -0,0 +1,236 @@
+
+replEvents :: KeyEvents REPLEvent
+replEvents = allKeyEvents $ \case
+  CancelRunningProgramEvent -> "cancel running program"
+  ToggleCustomKeyHandlingEvent -> "toggle custom key handling"
+  TogglePilotingModeEvent -> "toggle piloting mode"
+
+defaultReplBindings :: [(REPLEvent, [Binding])]
+defaultReplBindings = allBindings $ \case
+  CancelRunningProgramEvent -> [ctrl 'c', bind V.KEsc]
+  TogglePilotingModeEvent -> [meta 'p']
+  ToggleCustomKeyHandlingEvent -> [meta 'k']
+
+-- ----------------------------------------------
+--                 REPL EVENTS
+-- ----------------------------------------------
+
+data WorldEvent
+  = ViewBaseEvent
+  | ShowFpsEvent
+  | MoveViewEvent AbsoluteDir
+  deriving (Eq, Ord, Show)
+
+instance Enum WorldEvent where
+  fromEnum = \case
+    ViewBaseEvent -> 0
+    ShowFpsEvent -> 1
+    MoveViewEvent d -> 2 + fromEnum d
+  toEnum = \case
+    0 -> ViewBaseEvent
+    1 -> ShowFpsEvent
+    n -> MoveViewEvent . toEnum $ n - 2
+
+instance Bounded WorldEvent where
+  minBound = ViewBaseEvent
+  maxBound = MoveViewEvent maxBound
+
+worldPanelEvents :: KeyEvents WorldEvent
+worldPanelEvents = allKeyEvents $ \case
+  ViewBaseEvent -> "view base"
+  ShowFpsEvent -> "show fps"
+  MoveViewEvent d -> "move view " <> directionSyntax (DAbsolute d)
+
+defaultWorldPanelBindings :: [(WorldEvent, [Binding])]
+defaultWorldPanelBindings = allBindings $ \case
+  ViewBaseEvent -> [bind 'c']
+  ShowFpsEvent -> [bind 'f']
+  MoveViewEvent DWest -> [bind V.KLeft, bind 'h']
+  MoveViewEvent DSouth -> [bind V.KDown, bind 'j']
+  MoveViewEvent DNorth -> [bind V.KUp, bind 'k']
+  MoveViewEvent DEast -> [bind V.KRight, bind 'l']
+
+-- ----------------------------------------------
+--                 ROBOT EVENTS
+-- ----------------------------------------------
+
+data RobotEvent
+  = MakeEntityEvent
+  | ShowZeroInventoryEntitiesEvent
+  | CycleInventorySortEvent
+  | SwitchInventorySortDirection
+  | SearchInventoryEvent
+  deriving (Eq, Ord, Show, Enum, Bounded)
+
+robotPanelEvents :: KeyEvents RobotEvent
+robotPanelEvents = allKeyEvents $ \case
+  MakeEntityEvent -> "make entity"
+  ShowZeroInventoryEntitiesEvent -> "show zero inventory entities"
+  CycleInventorySortEvent -> "cycle inventory sort"
+  SwitchInventorySortDirection -> "switch inventory direction"
+  SearchInventoryEvent -> "search inventory"
+
+defaultRobotPanelBindings :: [(RobotEvent, [Binding])]
+defaultRobotPanelBindings = allBindings $ \case
+  MakeEntityEvent -> [bind 'm']
+  ShowZeroInventoryEntitiesEvent -> [bind '0']
+  CycleInventorySortEvent -> [bind ';']
+  SwitchInventorySortDirection -> [bind ':']
+  SearchInventoryEvent -> [bind '/']
+
+-- ----------------
+-- Helper methods
+
+allKeyEvents :: (Ord e, Bounded e, Enum e) => (e -> Text) -> KeyEvents e
+allKeyEvents f = keyEvents $ map (f &&& id) enumerate
+
+allBindings :: (Bounded e, Enum e) => (e -> [Binding]) -> [(e, [Binding])]
+allBindings f = map (\e -> (e, f e)) [minBound .. maxBound]
```

Comment:
Potentially:
```suggestion
allBindings :: (Bounded e, Enum e) => (e -> [Binding]) -> NonEmpty (e, [Binding])
allBindings f = NE.map (\e -> (e, f e)) enumerateNonEmpty
```
I've gone through the exercise of plumbing it through in #2013.

Reply:
@kostmo could you please explain how this will help us? Maybe I am missing something, but it seems like you prove the list has at least one element and then immediately discard that with `NE.toList`.

Here, `brick` is happy to take a list with any number of elements, and zero elements seem perfectly well-defined, so I do not see the point.

I did miss `enumerate` though, that definitely should be here.

Reply:
Agreed, the current Brick consumer does not require nonemptiness. I tend to "hold on to the proofs" as long as possible for the benefit of refactoring.  But this is sometimes in tension with the YAGNI principle.  So I leave it to you to decide.

Reply:
In this case, let's keep it simple. But thanks for showing the non-empty version; I will keep it in mind. 👍